### PR TITLE
[youtube] download live from the start, to the end

### PR DESCRIPTION
**Pull request by nao20010128nao**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

I have used ytarchive for reference, but the code is written from the scratch.

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This attempts to record YouTube live from the start. This does not provide general DASH live downloader, but some part of the code could be reused.

TODO

- [x] #1036, merge using ffmpeg after that
- [x] Work better for streams which does not have very early segments, found in some long-lasting live you can find by searching "jazz live"
- [x] do re-extraction every 5 hours, to prevent from being stopped when one manifest reached its time limit
- [ ] Fix `KeyboardInterrupt` on Windows (works on Linux, by @nao20010128nao)
- [x] Do fixup for duplicate moov atom

## Test cases
Long-lasting live: https://www.youtube.com/watch?v=fEvM-OUbaKs
Shorter live: find one here: https://nijisanji.net/ https://schedule.hololive.tv/

## Does it work on M3U8 formats? ~~Why do you enable `include_live_dash`?~~
YouTube doesn't provide older segments for m3u8 formats. Confirmed on my end, and ytarchive didn't that too.
~~The README says "These formats don't download properly" for `include_live_dash`, but enabling it along with `download_live_from_start` do download properly.~~

# USAGE
After downloading this PR, add `--live-from-start` to options to enable the feature.

e.g:
```bash
./yt-dlp.sh fEvM-OUbaKs --live-from-start
```
... and add any options you want

You always need this even after this PR is merged.

Edited by @pukkandan

### CODE REVIEW

**Comment by nao20010128nao** [comment]
> Support downloading 2 or more streams at once, merge using ffmpeg after that

Decided to make a separate PR for this, since it requires rewrite outside `YoutubeDlFromStartDashFD`

**Comment by pukkandan** [comment]
I am turning off notifications for this. ping me once its ready for review

**Comment by nao20010128nao** [comment]
okay

File: yt_dlp/downloader/youtube_dl_from_start.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    FD_NAME = 'ytlivestartdash'
+
+    @staticmethod
+    def _manifest_fragments(ie: YoutubeIE, mpd_url, stream_number, begin_index=0, fetch_span=5000):
+        known_idx = begin_index + 1
+        no_fragment_count = 0
+        prev_dl = time_millis()
+        while True:
+            if no_fragment_count > 5:
+                return
```
Comment (pukkandan) [inline yt_dlp/downloader/youtube_dl_from_start.py:30]: Is this reliable? Maybe we should let the user press ctrl+c to end stream

File: yt_dlp/downloader/youtube_dl_from_start.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+        stream_number = info_dict.get('manifest_stream_number', 0)
+        yie: YoutubeIE = self.ydl.get_info_extractor(YoutubeIE.ie_key())
+
+        download_start_time = ctx.get('start') or (time_millis() / 1000)
+        live_start_time = traverse_obj(root_info_dict, ('__original_infodict', 'release_timestamp'))
+        begin_index = 0
+        if live_start_time is not None and download_start_time - live_start_time > 432000:
+            self.report_warning('Starting downloading from recent 120 hours of live, because YouTube does not have the data. Please file an issue if you think it is wrong.', only_once=True)
+            begin_index = -86400  # 432000 / 5, where 5 is the duration of each segments
```
Comment (pukkandan) [inline yt_dlp/downloader/youtube_dl_from_start.py:79]: I think the fragment duration can be controlled by the uploader and is not constant

File: yt_dlp/downloader/youtube_dl_from_start.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    def _get_fragments(self, info_dict, root_info_dict, ctx):
+        manifest_url = info_dict.get('manifest_url')
+        if not manifest_url:
+            self.report_error('URL for MPD manifest is not known; there is a problem in YoutubeIE code')
+
+        stream_number = info_dict.get('manifest_stream_number', 0)
+        yie: YoutubeIE = self.ydl.get_info_extractor(YoutubeIE.ie_key())
+
+        download_start_time = ctx.get('start') or (time_millis() / 1000)
+        live_start_time = traverse_obj(root_info_dict, ('__original_infodict', 'release_timestamp'))
```
Comment (pukkandan) [inline yt_dlp/downloader/youtube_dl_from_start.py:75]: is this not present in the info_dict itself?

File: yt_dlp/downloader/youtube_dl_from_start.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            last_seq = int(re.search(r'(?:/|^)sq/(\d+)', fragments[-1]['path']).group(1)) + 1
+            if begin_index < 0 and known_idx <= 0:
+                # skip from the start when it's negative value
+                known_idx = last_seq + begin_index
+            for idx in range(known_idx, last_seq):
+                yield {
+                    'frag_index': idx - 1,
+                    'index': idx - 1,
+                    'url': urljoin(fragment_base_url, 'sq/%d' % idx),
+                }
```
Comment (pukkandan) [inline yt_dlp/downloader/youtube_dl_from_start.py:50]: As I understand, this URL fetching and the below max duration check are the only YouTube specific code. An I correct?

File: yt_dlp/downloader/youtube_dl_from_start.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            self.report_error('URL for MPD manifest is not known; there is a problem in YoutubeIE code')
+
+        stream_number = info_dict.get('manifest_stream_number', 0)
+        yie: YoutubeIE = self.ydl.get_info_extractor(YoutubeIE.ie_key())
+
+        download_start_time = ctx.get('start') or (time_millis() / 1000)
+        live_start_time = traverse_obj(root_info_dict, ('__original_infodict', 'release_timestamp'))
+        begin_index = 0
+        if live_start_time is not None and download_start_time - live_start_time > 432000:
+            self.report_warning('Starting downloading from recent 120 hours of live, because YouTube does not have the data. Please file an issue if you think it is wrong.', only_once=True)
```
Comment (pukkandan) [inline yt_dlp/downloader/youtube_dl_from_start.py:78]: see above comment

File: yt_dlp/downloader/youtube_dl_from_start.py (hunk 1 above)
Comment (nao20010128nao) [reply yt_dlp/downloader/youtube_dl_from_start.py:30]: this is for when the downloader can't find no new fragments. Ctrl+C handling is the responsible of FragmentFD, not here

File: yt_dlp/downloader/youtube_dl_from_start.py (hunk 2 above)
Comment (nao20010128nao) [reply yt_dlp/downloader/youtube_dl_from_start.py:79]: I don't think uploader can do this, but YouTube may. Will try to fix

File: yt_dlp/downloader/youtube_dl_from_start.py (hunk 3 above)
Comment (nao20010128nao) [reply yt_dlp/downloader/youtube_dl_from_start.py:75]: which is given to FD?

File: yt_dlp/downloader/youtube_dl_from_start.py (hunk 4 above)
Comment (nao20010128nao) [reply yt_dlp/downloader/youtube_dl_from_start.py:50]: yes. YouTube-specific code.

File: yt_dlp/downloader/youtube_dl_from_start.py (hunk 5 above)
Comment (nao20010128nao) [reply yt_dlp/downloader/youtube_dl_from_start.py:78]: answered

File: yt_dlp/downloader/youtube_dl_from_start.py (hunk 2 above)
Comment (nao20010128nao) [reply yt_dlp/downloader/youtube_dl_from_start.py:79]: so maybe I need to change it to `lack_early` or something else, and let `_manifest_fragments` calculate fragment sequence to start from by itself? :thinking:

Comment (nao20010128nao) [reply yt_dlp/downloader/youtube_dl_from_start.py]: dome for this

**Comment by Rikk** [comment]
Just found this PR. I was trying to download a live starting from some minutes before current time. Is there much work left on this?

**Comment by pukkandan** [comment]
If you clone this branch, it should work. The delay is mainly because I have been busy. I will try to get this in the next release

**Comment by nao20010128nao** [comment]
Do I have to wait for #1125 to be merged?

**Comment by pukkandan** [comment]
> Would #1125 conflict with this PR?

Since this PR doesnt interact directly with minicurses, I don't think so

**Comment by nao20010128nao** [comment]
okay

File: yt_dlp/downloader/youtube_dl_from_start.py (hunk 1 above)
Comment (nao20010128nao) [reply yt_dlp/downloader/youtube_dl_from_start.py:30]: tl;dr unreliable. requires to get header of **any segment**
https://github.com/Kethsar/ytarchive/blob/master/ytarchive.py#L1090

**Comment by nao20010128nao** [comment]
Copying my message from Discord for other contributors wanting to support native live downloaders:

> when reusing the code for other live stream formats, you must do that:
> 1. copy _calculate_fragment_count from YoutubeDlFromStartDashFD
this is needed to let FFD know it's live stream, so it must be common for native livestream downloader
> 2. make _get_fragments return iterator, instead of list
use generator as I do in #888
> 3. modify YoutubeDL appropriately
>
> then everything is fine to go, and the same for ad skipping for twitch

**Comment by nao20010128nao** [comment]
```diff
diff --git a/yt_dlp/downloader/fragment.py b/yt_dlp/downloader/fragment.py
index 3d66ae19d..b5e50496b 100644
--- a/yt_dlp/downloader/fragment.py
+++ b/yt_dlp/downloader/fragment.py
@@ -4,6 +4,8 @@ import os
 import time
 import json
 import http.client
+import threading
+import signal
 from math import ceil
 
 try:
@@ -400,15 +402,28 @@ class FragmentFD(FileDownloader):
             job = tpe.submit(thread_func, idx, ctx, fragments, info_dict, tpe)
             spins.append((tpe, job))
 
+        event = threading.Event()
+        def _signal_handler(signo, _frame):
+            event.set()
+        
+        signal.signal(signal.SIGINT, _signal_handler)
+        signal.signal(signal.SIGTERM, _signal_handler)
+
         result, kint = True, None
         for tpe, job in spins:
-            try:
-                result = result and job.result()
-            except KeyboardInterrupt as ex:
-                interrupt_trigger[0] = False
-                kint = ex
-            finally:
-                tpe.shutdown(wait=True)
+            while job.running() and not event.is_set():
+                try:
+                    event.wait(10)
+                    result = result and job.result(0.1)
+                    break
+                except concurrent.futures.TimeoutError:
+                    continue
+                except KeyboardInterrupt as ex:
+                    interrupt_trigger[0] = False
+                    kint = ex
+                    break
+                finally:
+                    tpe.shutdown(wait=True)
         if not interrupt_trigger[0] and not any(arg[0].get('live') for arg in args):
             # rethrow if it's not live
             raise kint or KeyboardInterrupt()
diff --git a/yt_dlp/extractor/youtube.py b/yt_dlp/extractor/youtube.py
index 0db49ace8..984fe57e5 100644
--- a/yt_dlp/extractor/youtube.py
+++ b/yt_dlp/extractor/youtube.py
@@ -1729,7 +1729,7 @@ class YoutubeIE(YoutubeBaseInfoExtractor):
                 try:
                     fmts, _ = self._extract_mpd_formats_and_subtitles(
                         mpd_url, None, note=False, errnote=False, fatal=False)
-                except BaseException:
+                except ExtractorError:
                     fmts = None
                 if not fmts:
                     no_fragment_score += 1
@@ -1745,7 +1745,7 @@ class YoutubeIE(YoutubeBaseInfoExtractor):
                 try:
                     urlh = self._request_webpage(
                         last_segment_url, None, note=False, errnote=False, fatal=False)
-                except BaseException:
+                except ExtractorError:
                     urlh = None
                 if not urlh:
                     no_fragment_score += 1
```
looks like Python on Windows completely refuses Ctrl+C while waiting for threads :confused:

File: yt_dlp/extractor/youtube.py [hunk 6]
```diff
@@ -1738,9 +1737,8 @@ def mpd_feed(format_id):
             """
             @returns (manifest_url, manifest_stream_number, is_live) or None
             """
-            if time.time() > expiration_time:
-                with lock:
-                    refetch_manifest(format_id)
+            with lock:
```
Comment (nao20010128nao) [inline yt_dlp/extractor/youtube.py:1740]: wouldn't this lock unnecessarily many times? `mpd_feed` is called every time a fragment yielded.

Comment (pukkandan) [reply yt_dlp/extractor/youtube.py:1740]: it will lock, check time and unlock. Is that an issue? If so, this can be reverted

Comment (nao20010128nao) [reply yt_dlp/extractor/youtube.py:1740]: I intentionally duplicated code to reduce number of entering locks, as it may wait for other threads

Comment (nao20010128nao) [reply yt_dlp/extractor/youtube.py:1740]: decided not to revert this as the effect caused by unnecessary locks seems relatively small

File: yt_dlp/downloader/dash.py [hunk 7]
```diff
@@ -1,9 +1,12 @@
 from __future__ import unicode_literals
+import time
 
 from ..downloader import get_suitable_downloader
 from .fragment import FragmentFD
 
-from ..utils import urljoin
+from ..utils import (
+    urljoin,
+)
```
Comment (pukkandan) [inline yt_dlp/downloader/dash.py:9]:
```suggestion
from ..utils import rljoin
```

**Comment by Rikk** [comment]
Very nice that this went live!
It seems it doesn't work when a format is specified (-f), this info is missing in Readme.
How to stop downloading properly now that ctrl+c doesn't work? ("Ctrl+C does not work on Windows when used with parallel threads") I suppose closing prompt window will leave files incomplete/improperly closed, right?

**Comment by nao20010128nao** [comment]
@Rikk 
> It seems it doesn't work when a format is specified (-f), 

It should work. What format are you specifying?

> How to stop downloading properly now that ctrl+c doesn't work?

I assume you're using Windows, the only way to stop it is to wait for the live ends, or being made private.

> I suppose closing prompt window will leave files incomplete/improperly closed, right?

Yes, but directly feeding these files to ffmpeg should work.

**Comment by paradix** [comment]
> Yes, but directly feeding these files to ffmpeg should work.

How would you achieve that?

**Comment by nao20010128nao** [comment]
@paradix Like this:

```
ffmpmeg -i videopart.mp4.part -i audiopart.mp4.part -c copy merged.mp4
```

yt-dlp uses more complex command than this, but fundamentally same at least for merging the two from youtube.

**Comment by Rikk** [comment]
@nao20010128nao 
> > It seems it doesn't work when a format is specified (-f),
> 
> It should work. What format are you specifying?
>
The problem is that I was specifying the formats displayed by **`--list-formats`**, for example:
formats
Any of these will result in `ERROR: [youtube] <vid-id>: Requested format is not available`
When no format is specified (or a good one is used) it works properly, eg: `[info] <vid-id>: Downloading 1 format(s): 248+140`

Therefore, I think the bug is the **`--list-formats`** command returning above strange formats for live videos.

**Comment by pukkandan** [comment]
Those are formats that are downloaded "normally". Use `--list-formats --live-from-start` to list the formats that you can download with live-from-start

**Comment by CeruleanSky** [comment]
> > How to stop downloading properly now that ctrl+c doesn't work?
> 
> I assume you're using Windows, the only way to stop it is to wait for the live ends, or being made private.

Ctrl+Break will work. Some devices might not have the key, so you can use the On-Screen keyboard to do so.  Not quite sure about the properness of it, if you need special handling [signal.SIGBREAK](https://docs.python.org/3/library/signal.html#signal.SIGBREAK) seems to be available from windows.

**Comment by grichecth** [comment]
@Lesmiscore Sorry to bother you, but is there any way your code that handles the re-extraction every 5 hours could be implemented when using --no-live-from-start? This is an open issue, https://github.com/yt-dlp/yt-dlp/issues/1998. I'd be willing to pay $50 usd for it to be fixed if you are able. I use this for a personal project and being limited by 6 hours is messing up what I'm trying to achieve. Thanks either way!

**Pull request merged**